Fix #5 Add option to remove unused imports

tommythorsen opened the pull request:
The algorithm for calculating whether an import is unused, is really
simple. It only looks at the body of the file, and checks if the last
segment of the import path (typically a class name or a static function
name) is refernenced in the code. This makes us unable to remove imports
with wildcards, but apart from that, it seems to work well.

The "removeUnused" configuration parameter is false by default.

tommythorsen commented:
Hi!

Thanks for making a great plugin!

I hope you will consider including my pull request, as this is something I really need for a few projects at my workplace.

I have tested this on a fairly large java repository, and it is able to remove a lot of unused imports without causing any problems.

If there is something you would like changed, or implemented differently, don't hesitate to tell me.

tommythorsen commented:
(Related to issue #5)

ctubbsii commented:
@tommythorsen Thanks for the contribution. If I understand your description, it seems like this would fail safe when encountering false positives... only removing imports which are *definitely* not in use, but possibly leaving some in place when they aren't needed. I think that's a good solution.

I'll need to take some time to look at the code, but if everything checks out, I'd be happy to merge it. It's better than nothing. A few considerations I'll be looking for:

1. Affect on performance
2. How failures are handled (including how broken code, such as that described in javaparser/javaparser#1433 is affected)
3. User experience (configuration)

Have you had a chance to examine performance already?

tommythorsen commented:
1. I have run it on a decently sized codebase, and didn't notice any horrible slowdown, but I have not made any actual measurements. I am introducing more code, that looks at the entire contents of every file, so I would expect some slowdown.

However, when this feature is turned off (by leaving the `removeUnused` config parameter to false), there should not be any slowdown at all. This means that people who are not interested in this new functionality should not be affected at all. Only those who are interested in pruning unused imports can turn this feature on, and get some extra functionality, at the cost of burning some more CPU cycles.

2. This I don't know. I am only looking at low level tokens, and not more high level things, so I would think there wouldn't be a problem. A java file would have to be in pretty bad shape before the tokenizer would stop functioning.

3. There is a configuration parameter called `removeUnused`. I have added it in a way similar to other configuration parameters. You might have to add a short sentence about it over at http://code.revelc.net/impsort-maven-plugin/, though.

kpm1985 commented:
Backseat driver: have not looked at the algorithms but if you're super
concerned abouy performance one option you could tune the parser. Sometimes
you can beat a regex. (Might be irrelevant comes from compiler experience
parsing source).

On Mon, Apr 9, 2018, 1:40 PM Tommy Thorsen <notifications@github.com> wrote:

>
>    1. I have run it on a decently sized codebase, and didn't notice any
>    horrible slowdown, but I have not made any actual measurements. I am
>    introducing more code, that looks at the entire contents of every file, so
>    I would expect some slowdown.
>
> However, when this feature is turned off (by leaving the removeUnused
> config parameter to false), there should not be any slowdown at all. This
> means that people who are not interested in this new functionality should
> not be affected at all. Only those who are interested in pruning unused
> imports can turn this feature on, and get some extra functionality, at the
> cost of burning some more CPU cycles.
>
>    1.
>
>    This I don't know. I am only looking at low level tokens, and not more
>    high level things, so I would think there wouldn't be a problem. A java
>    file would have to be in pretty bad shape before the tokenizer would stop
>    functioning.
>    2.
>
>    There is a configuration parameter called removeUnused. I have added
>    it in a way similar to other configuration parameters. You might have to
>    add a short sentence about it over at
>    http://code.revelc.net/impsort-maven-plugin/, though.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/revelc/impsort-maven-plugin/pull/6#issuecomment-379887072>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Acg-LE6iJnCKCX-CeKZlXIPYLKDHHvKwks5tm8dQgaJpZM4TMAzo>
> .
>

ctubbsii commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+        JavaToken token = tokenRange.get().getBegin();
+        while (token != null && token != JavaToken.INVALID) {
+          String tokenString = token.toString();
+          if (!tokenString.isEmpty() && Character.isLetter(tokenString.charAt(0))) {
```
Should use `Character.isJavaIdentifierStart` instead of `Characteri.isLetter` (see https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8)

tommythorsen replied:
I was not aware of `Character.isJavaIdentifierStart`, but that fits perfectly. Thanks!

ctubbsii commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+        while (token != null && token != JavaToken.INVALID) {
+          String tokenString = token.toString();
+          if (!tokenString.isEmpty() && Character.isLetter(tokenString.charAt(0))) {
+            tokens.add(token.toString());
```
The toString here is redundant. Just use `tokenString` created 2 lines previous.

tommythorsen replied:
Good catch. Will fix!

ctubbsii commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+          if (!tokenString.isEmpty() && Character.isLetter(tokenString.charAt(0))) {
+            tokens.add(token.toString());
+          }
+          Optional<JavaToken> next = token.getNextToken();
```
Is there a `hasNextToken` to loop on, in the while condition instead of pre-loading the token at the end of the previous loop?

tommythorsen replied:
Unfortunately, no. But if I could convince you to move to a newer version of JavaParser, the TokenRange class gets an `Iterator<Token> iterator()` method, which would really simplify this code.

ctubbsii replied:
Oh, I haven't been watching for updates. I think I used the latest that was available at the time. Currently, it is up to version 3.6.0. As long as it doesn't break anything and doesn't require a JDK newer than 8, feel free to bump it as needed in this pull request to simplify the parse code.

ctubbsii commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+  private static Set<String> extractTokens(CompilationUnit unit) {
+    Set<String> tokens = new LinkedHashSet<String>();
+    NodeList<TypeDeclaration<?>> types = unit.getTypes();
+    types.forEach((Node node) -> {
```
Out of curiosity, why was the explicit "Node" type necessary here in this lambda?

matozoid replied:
I'll make a wild guess that `<?>` was throwing the type inference off.

tommythorsen replied:
Actually, I tested without the explicit type, and it works. I'm not sure why I added it, but it was probably because I thought it wouldn't work with the `<?>` otherwise. I can remove this type.

ctubbsii replied:
It doesn't matter to me. I was just curious. I have only just begun to learn all the quirks of lambdas in Java :smiley_cat:

ctubbsii commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+   */
+  private static void removeUnusedImports(Set<Import> imports, Set<String> tokens) {
+    imports.removeIf(i -> {
+      String[] segments = i.getImport().split("\\.");
```
Rather than the double escape for `.`, I tend to prefer `[.]`. It reduces confusion, especially in Java, by avoiding unnecessary double-escaping for regex.

tommythorsen replied:
Will change.

ctubbsii commented on src/test/java/net/revelc/code/impsort/ImpSortTest.java:
```diff
@@ -30,7 +34,27 @@
+  }
+
+  @Test
+  public void testUnused() throws IOException {
```
Nice test!

tommythorsen replied:
Thanks :)

ctubbsii commented on src/test/java/net/revelc/code/impsort/ImpSortTest.java:
```diff
@@ -30,7 +34,27 @@
   public void testSort() throws IOException {
     Path p = Paths.get(System.getProperty("user.dir"), "src", "it", "plugin-test", "src", "test", "java", "net", "revelc", "code", "imp", "PluginIT.java");
-    new ImpSort(eclipseDefaults).parseFile(p);
+    new ImpSort(eclipseDefaults, false).parseFile(p);
```
I'm wondering if the functionality should be added to the Grouper instead of directly in the ImpSort class, since the Grouper tends to handle most of the other configuration options. It would save from adding a boolean flag as a parameter, which I think tends to be not very elegant. (See https://stackoverflow.com/questions/6107221/alternatives-to-passing-a-flag-into-a-method for some other alternatives to the boolean, if Grouper is not a good fit.)

tommythorsen replied:
Some observation on this:
1. The `extractTokens` function that walks through the `CompilationUnit` and extracts the tokens, should be in the `ImpSort` class. The alternative would be to pass the `CompilationUnit` to the Grouper, and let the Grouper extract the tokens, but this feels wrong.
2. Since we only want to run `extractTokens` if `removeUnused` is `true`, the `removeUnused` flag must be available for the `ImpSort` class to check.

I think I'll take some time to consider the alternatives to the boolean, as I think it is hard to move this functionality out of ImpSort in a straight-forward way.

ctubbsii replied:
Changes to this could be deferred to a future design change. I wouldn't hold up merging on the this comment. Just thought it was worth asking. :smiley_cat:

tommythorsen commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+  private static void removeUnusedImports(Set<Import> imports, Set<String> tokens) {
+    imports.removeIf(i -> {
+      String[] segments = i.getImport().split("\\.");
+      if (segments.length == 0) return false;
```
Wouldn't it be nice to handle invalid code graciously as well? I can certainly remove this check if you really want me to, but it just feels like good defensive programming to leave it in.

ctubbsii replied:
Perhaps it's fine, but it looks like dead code to me, since it can never happen. Invalid code which would cause a compilation failure would fail long before this stage, within the parser itself. However, it's possible a newer version of the parser will allow partial processing of invalid code.  To protect against that case, instead of returning false, I would suggest `throw new AssertionError("Parse tree includes invalid import statements")`

tommythorsen commented on src/main/java/net/revelc/code/impsort/ImpSort.java:
```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
+   * to figure out whether or not an import is unused.
+   */
+  private static Set<String> extractTokens(CompilationUnit unit) {
+    Set<String> tokens = new LinkedHashSet<String>();
```
The order does not matter, and there is no good reason to use `LinkedHashSet`. I'll change this.

tommythorsen commented on src/main/java/net/revelc/code/impsort/maven/plugin/AbstractImpSortMojo.java:
```diff
@@ -132,6 +132,14 @@
+  /**
+   * Configures whether to remove unused imports.
+   *
+   * @since 1.0.2
```
I'll fix this.

tommythorsen commented:
I added a fixup commit that addresses all review comments, except for the last one. I will address the last comment in a separate commit. I will also look into creating a performance test.

tommythorsen commented:
I added a commit that removes the unwanted boolean parameter: https://github.com/revelc/impsort-maven-plugin/pull/6/commits/fad9407fac6b82c5db1419755b743daa1e6e46f0

It's probably more brutal than you were hoping for, but it's the sanest solution I could find to the problem. The solution itself is described in more detail in the commit message.

If you don't like this refactoring, just say so, and I'll roll it back, but in that case, I would really like some pointers from you as to what kind of solution you would prefer.

ctubbsii commented on pom.xml:
```diff
@@ -90,7 +90,7 @@
         <groupId>com.github.javaparser</groupId>
         <artifactId>javaparser-core</artifactId>
-        <version>3.2.12</version>
+        <version>3.5.0</version>
```
Why not the latest, 3.6.0?

ctubbsii commented:
Commit 3d7f03578857a0f080ddf66c951adb469c1b51a2 looks good to me. I'm going to merge that one in.

ctubbsii commented:
Okay, so I just noticed a big problem: javadocs. The current detection removes imports used by javadoc `@see` and `@link` tags. I could still merge it in... but it's not as useful of a feature if it removes those.

ctubbsii commented:
Also, I ran against a large codebase, and it was only a few tenths of a second different, so performance isn't an issue.

tommythorsen commented:
Good to hear the performance is ok. I can take a look at the javadoc issue. Do you have some examples of usages that fail?

ctubbsii commented:
I've merged a squashed version of your commits (conflicts resolved) as 66732f7d1cfc4eada82e3ea1c54dae9dc9a66e49 up through commit 3d7f03578857a0f080ddf66c951adb469c1b51a2, but not including the refactor in fad9407fac6b82c5db1419755b743daa1e6e46f0.

I will create a separate issue for the removal of imports used by javadoc. That way, we can make incremental progress.
